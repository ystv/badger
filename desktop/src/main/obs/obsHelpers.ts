import {
  castMediaSourceSettings,
  obsConnection,
  OBSVideoSettings,
  SceneItem,
} from "./obs";
import invariant from "../../common/invariant";
import type { Media } from "@badger/prisma/types";
import { getLogger } from "../base/logging";
import { getLocalMedia } from "../media/mediaManagement";

const logger = getLogger("obsHelpers");

/*
 * This file contains helper functions that wrap obsConnection in obs.ts.
 */

export type MediaType = Media & {
  containerType: "rundownItem" | "continuityItem" | "asset";
  containerId: number;
  containerName: string;
  order: number;
};

const MEDIA_SOURCE_PREFIX = "Badger Media ";
export const OLD_CONTINUITY_SCENE_NAME_REGEXP = /^\d+ - .+? \[#(\d+)]$/;
export const NEW_SCENE_NAME_REGEXP =
  /^\d - .*? \[(VT|Continuity|Asset) #(\d+)\]$/;

/**
 * Adds a file to an OBS scene as a media source, creating the scene if it does not exist.
 * @param info the media to add
 * @param replaceMode how to handle OBS scenes that already exist:
 *   "none": do not replace existing sources
 *   "replace": replace pre-existing Badger scenes, but not scenes containing other sources
 *   "force": replace the contents of the scene no matter what
 * @return
 *   done: true if the media was added to the scene
 *   warnings: any warnings that occurred
 *   promptReplace: if replaceMode is *not* "replace" or "force", and the scene already exists, this will be "replace" or "force" to indicate that the user should be prompted to replace the scene
 */
export async function addOrReplaceMediaAsScene(
  info: MediaType,
  replaceMode: "none" | "replace" | "force",
): Promise<{
  done: boolean;
  warnings: string[];
  promptReplace?: "replace" | "force";
}> {
  const warnings: string[] = [];
  const warn = (v: string) => {
    logger.warn("addMediaAsScene:", v);
    warnings.push(v);
  };

  invariant(obsConnection, "no OBS connection");
  const localMedia = getLocalMedia();
  const item = localMedia.find((x) => x.mediaID === info.id);
  invariant(
    item !== undefined,
    `No local media for media ${info.id} [${info.name}] in addMediaAsScene`,
  );

  const videoSettings = await obsConnection.getVideoSettings();

  const mediaSourceName = MEDIA_SOURCE_PREFIX + info.id.toString(10);

  const containerTypeStr =
    info.containerType === "rundownItem"
      ? "VT"
      : info.containerType === "continuityItem"
        ? "Continuity"
        : "Asset";
  const sceneTitle = `${info.order} - ${info.containerName} [${containerTypeStr} #${info.containerId}]`;
  // Sanity checks
  invariant(
    mediaSourceName.startsWith(MEDIA_SOURCE_PREFIX),
    "Generated media source name that won't match our prefix checks later: " +
      mediaSourceName,
  );
  invariant(
    sceneTitle.match(NEW_SCENE_NAME_REGEXP),
    "Generated scene title that won't match our regex checks later: " +
      sceneTitle,
  );

  const scenes = await obsConnection.listScenes();
  const ours = scenes.find((x) => x.sceneName === sceneTitle);
  if (!ours) {
    // The scene doesn't exist, create it from scratch.
    logger.debug("addMediaAsScene: creating scene", sceneTitle);
    await obsConnection.createScene(sceneTitle);
    await _doAddMediaToScene(sceneTitle, mediaSourceName, item, videoSettings);
    return { warnings, done: true };
  }

  // The scene exists. Check if it has the source we want, and if not, add it.
  logger.debug("addMediaAsScene: found scene", ours);
  const items = await obsConnection.getSceneItems(sceneTitle);
  // If the scene is empty, we can go ahead and add our source
  if (items.length === 0) {
    logger.debug(
      "addMediaAsScene: existing scene doesn't have our source. Adding.",
    );
    await _doAddMediaToScene(sceneTitle, mediaSourceName, item, videoSettings);
    return { warnings, done: true };
  }

  // If the scene is non-empty, but only has our source, we're probably fine - do some sanity checks.
  if (items.length === 1 && items[0].sourceName === mediaSourceName) {
    // The scene exists, and it has a source that matches our naming convention. Check if it's the same file.
    logger.debug("addMediaAsScene: found existing source");
    const settings = await obsConnection.getSourceSettings(mediaSourceName);
    if (!castMediaSourceSettings(settings)) {
      // Should never happen unless the user manually creates a different source, or there's a Badger version incompatibility.
      if (replaceMode === "force") {
        warn("Existing source is not a media source. Forcing replacement.");
        await obsConnection.removeSceneItem(sceneTitle, items[0].sceneItemId);
        await _doAddMediaToScene(
          sceneTitle,
          mediaSourceName,
          item,
          videoSettings,
        );
        return { warnings, done: true };
      } else {
        warn("Existing source is not a media source.");
        return { warnings, done: false };
      }
    }
    if (settings.inputSettings.local_file === item.path) {
      logger.debug(
        "addMediaAsScene: existing source is the same file. Nothing to do.",
      );
      return { warnings, done: false };
    }

    // The scene exists, and has a source that matches our naming convention, but it's a different file.
    // This should never actually happen. When a media file is replaced, we create a new Media object with a new ID,
    // which would mean the source has a new name (see the definition of mediaSourceName). There being a source with the
    // same ID and name but different path suggests that the user has manually changed it, and we shouldn't touch it.
    if (replaceMode === "force") {
      warn(
        `Media source ${mediaSourceName} in scene ${sceneTitle} has a different path (${settings.inputSettings.local_file}) than expected (${item.path}). Forcing replacement.`,
      );
      await obsConnection.replaceMediaSourceInScene(
        sceneTitle,
        mediaSourceName,
        item.path,
      );
      return { warnings, done: true };
    } else {
      warn(
        `Media source ${mediaSourceName} in scene ${sceneTitle} has a different path (${settings.inputSettings.local_file}) than expected (${item.path}). Not updating.`,
      );
      return {
        warnings,
        done: false,
        promptReplace: "force",
      };
    }
  }
  // If the scene is non-empty, but has only one other source, and it's a Badger source, we can replace it if the user permits it.
  if (
    items.length === 1 &&
    items[0].sourceName.startsWith(MEDIA_SOURCE_PREFIX)
  ) {
    if (replaceMode === "replace" || replaceMode === "force") {
      logger.debug(
        "addMediaAsScene: existing scene has one Badger source with mismatching ID. Replacing.",
      );
      await obsConnection.removeSceneItem(sceneTitle, items[0].sceneItemId);
      await _doAddMediaToScene(
        sceneTitle,
        mediaSourceName,
        item,
        videoSettings,
      );
      return { warnings, done: true };
    } else {
      warn(
        `Scene ${sceneTitle} has a pre-existing Badger source for a different media file.`,
      );
      return { warnings, done: false, promptReplace: "replace" };
    }
  }
  // The scene has non-Badger sources in it.
  if (replaceMode === "force") {
    // If the user has chosen to force replacement, we can just remove all the other sources and override the path for ours.
    // Note that we don't remove ours due to a potential race condition when removing and adding the same source.
    logger.debug(
      "addMediaAsScene: existing scene has non-Badger sources. Replacing as the replaceMode is 'force'.",
    );
    let oursPresent = false;
    for (const item of items) {
      if (item.sourceName == mediaSourceName) {
        oursPresent = true;
      } else {
        await obsConnection.removeSceneItem(sceneTitle, item.sceneItemId);
      }
    }
    if (oursPresent) {
      await obsConnection.replaceMediaSourceInScene(
        sceneTitle,
        mediaSourceName,
        item.path,
      );
    } else {
      await _doAddMediaToScene(
        sceneTitle,
        mediaSourceName,
        item,
        videoSettings,
      );
    }
    return { warnings, done: true };
  } else {
    // Bail.
    warn(
      `Scene ${sceneTitle} has non-Badger sources in it. Cowardly refusing to overwrite.`,
    );
    return { warnings, done: false, promptReplace: "force" };
  }
}

async function _doAddMediaToScene(
  sceneTitle: string,
  mediaSourceName: string,
  localMedia: { path: string },
  videoSettings: OBSVideoSettings,
) {
  invariant(obsConnection, "no OBS connection");
  const id = await obsConnection.createMediaSourceInput(
    sceneTitle,
    mediaSourceName,
    localMedia.path,
  );
  await obsConnection.setSceneItemTransform(sceneTitle, id, {
    boundsWidth: videoSettings.baseWidth,
    boundsHeight: videoSettings.baseHeight,
    boundsType: "OBS_BOUNDS_SCALE_INNER",
  });
}

export async function findScenes(): Promise<
  Array<{
    sceneName: string;
    type: "rundownItem" | "continuityItem" | "asset";
    itemId: number;
    sources: (SceneItem & {
      mediaID?: number;
    })[];
  }>
> {
  invariant(obsConnection, "no OBS connection");
  const scenes = await obsConnection.listScenes();
  const continuityScenes = scenes.filter(
    (x) =>
      OLD_CONTINUITY_SCENE_NAME_REGEXP.test(x.sceneName) ||
      NEW_SCENE_NAME_REGEXP.test(x.sceneName),
  );
  const sceneItems = await Promise.all(
    continuityScenes.map((x) => obsConnection!.getSceneItems(x.sceneName)),
  );
  return continuityScenes.map((x, i) => {
    const sources = sceneItems[i].map((item) => {
      if (!item.sourceName.startsWith(MEDIA_SOURCE_PREFIX)) {
        return item;
      }
      return {
        ...item,
        mediaID: parseInt(
          item.sourceName.slice(MEDIA_SOURCE_PREFIX.length),
          10,
        ),
      };
    });
    const oldRes = OLD_CONTINUITY_SCENE_NAME_REGEXP.exec(x.sceneName);
    if (oldRes) {
      return {
        sceneName: x.sceneName,
        type: "rundownItem",
        itemId: parseInt(oldRes[1], 10),
        sources,
      };
    }
    const newRes = NEW_SCENE_NAME_REGEXP.exec(x.sceneName);
    if (newRes) {
      return {
        sceneName: x.sceneName,
        type:
          newRes[1] === "VT"
            ? "rundownItem"
            : newRes[1] === "Continuity"
              ? "continuityItem"
              : "asset",
        itemId: parseInt(newRes[2], 10),
        sources,
      };
    }
    invariant(false, "Scene name didn't match either regex");
  });
}
