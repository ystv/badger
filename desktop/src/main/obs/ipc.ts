import { z } from "zod";
import { proc, r } from "../base/ipcRouter";
import { createOBSConnection, obsConnection } from "./obs";
import { addOrReplaceMediaAsScene, findScenes } from "./obsHelpers";
import { getLogger } from "loglevel";
import { serverAPI } from "../base/serverApiClient";
import invariant from "../../common/invariant";
import { selectedShow } from "../base/selectedShow";
import { getDevToolsConfig, getOBSSettings } from "../base/settings";
import { TRPCError } from "@trpc/server";
import { getLocalMedia } from "../media/mediaManagement";
import { Media } from "@badger/prisma/types";

const logger = getLogger("obs/ipc");

export const obsRouter = r({
  getConnectionState: proc
    .output(
      z.object({
        connected: z.boolean(),
        version: z.string().optional(),
        platform: z.string().optional(),
        error: z.string().optional(),
        availableRequests: z.array(z.string()).optional(),
        loadContinuityItems: z.boolean().optional(),
        loadRundownItems: z.boolean().optional(),
        loadAssets: z.boolean().optional(),
      }),
    )
    .query(async () => {
      // TODO[BDGR-136]: don't use the connection for this
      if (obsConnection === null) {
        return { connected: false };
      }
      try {
        const [version, settings] = await Promise.all([
          obsConnection.ping(),
          getOBSSettings(),
        ]);
        invariant(settings, "connected to OBS without settings");
        return {
          connected: true,
          version: version.obsVersion,
          platform: version.platformDescription,
          availableRequests: version.availableRequests,
        };
      } catch (e) {
        logger.warn("OBS connection error", e);
        return { connected: false, error: String(e) };
      }
    }),
  connect: proc
    .input(
      z.object({
        host: z.string(),
        port: z.coerce.number(),
        password: z.string(),
      }),
    )
    .mutation(async ({ input }) => {
      await createOBSConnection(input.host, input.password, input.port);
    }),
  addMediaAsScene: proc
    .input(
      z.object({
        containerType: z.enum(["rundownItem", "continuityItem", "asset"]),
        containerId: z.number(),
        rundownId: z.number().optional(),
        replaceMode: z.enum(["none", "replace", "force"]).default("none"),
      }),
    )
    .output(
      z.object({
        done: z.boolean(),
        warnings: z.array(z.string()),
        promptReplace: z.enum(["replace", "force"]).optional(),
      }),
    )
    .mutation(async ({ input }) => {
      const show = selectedShow.value;
      invariant(show, "No show selected");
      let container;
      switch (input.containerType) {
        case "rundownItem":
          invariant(input.rundownId, "rundownId required for rundownItem");
          container = show.rundowns
            .find((x) => x.id === input.rundownId)
            ?.items?.find((x) => x.id === input.containerId);
          break;
        case "continuityItem":
          container = show.continuityItems.find(
            (x) => x.id === input.containerId,
          );
          break;
        case "asset":
          invariant(input.rundownId, "rundownId required for rundownItem");
          container = show.rundowns
            .find((x) => x.id === input.rundownId)
            ?.assets?.find((x) => x.id === input.containerId);
          break;
      }
      invariant(container, "item not found");
      invariant(container.mediaId, "tried to add item with no media");
      const info = await serverAPI().media.get.query({ id: container.mediaId });
      invariant(
        info.continuityItems.length > 0,
        "obs.addMediaAsScene: No continuity item for media in obs.addMediaAsScene",
      );
      return await addOrReplaceMediaAsScene(
        {
          ...info,
          containerType: input.containerType,
          containerId: input.containerId,
          containerName: container.name,
          order: container.order,
        },
        input.replaceMode,
      );
    }),
  bulkAddMedia: proc
    .input(
      z.object({
        source: z.discriminatedUnion("type", [
          z.object({ type: z.literal("rundownItems"), rundownID: z.number() }),
          z.object({
            type: z.literal("rundownAssets"),
            rundownID: z.number(),
            category: z.string(),
          }),
          z.object({ type: z.literal("continuityItems") }),
        ]),
      }),
    )
    .mutation(async ({ input }) => {
      const show = selectedShow.value;
      invariant(show, "No show selected");
      let sources: Array<{
        media: Media | null;
        name: string;
        id: number;
        order: number;
      }>;
      const source = input.source;
      switch (source.type) {
        case "rundownItems": {
          invariant(source.rundownID, "Rundown ID required");
          const rundown = show.rundowns.find((x) => x.id === source.rundownID);
          invariant(rundown, `Rundown ${source.rundownID} not found`);
          sources = rundown.items.sort((a, b) => a.order - b.order);
          break;
        }
        case "rundownAssets": {
          invariant(source.rundownID, "Rundown ID required");
          const rundown = show.rundowns.find((x) => x.id === source.rundownID);
          invariant(rundown, `Rundown ${source.rundownID} not found`);
          sources = rundown.assets
            .filter((x) => x.category === source.category)
            .sort((a, b) => a.order - b.order);
          break;
        }
        case "continuityItems": {
          sources = show.continuityItems.sort((a, b) => a.order - b.order);
          break;
        }
        default:
          invariant(false, "Unknown source type");
      }
      if (sources.some((x) => !x.media || x.media.state !== "Ready")) {
        throw new TRPCError({
          code: "PRECONDITION_FAILED",
          message: "Not all media is ready",
        });
      }
      const locals = getLocalMedia();
      for (const source of sources) {
        if (
          source.media &&
          locals.some((x) => x.mediaID === source.media!.id)
        ) {
          await addOrReplaceMediaAsScene(
            {
              ...source.media,
              containerType: input.source.type.replace(/s$/, "") as
                | "rundownItem"
                | "continuityItem"
                | "asset",
              containerId: source.id,
              containerName: source.name,
              order: source.order,
            },
            "replace",
          );
        }
      }
    }),
  listBadgerScenes: proc
    .output(
      z.array(
        z.object({
          sceneName: z.string(),
          type: z.enum(["rundownItem", "continuityItem", "asset"]),
          itemId: z.number(),
          sources: z.array(
            z.object({
              mediaID: z.number().optional(),
            }),
          ),
        }),
      ),
    )
    .query(async () => {
      return await findScenes();
    }),
  dev: r({
    callArbitrary: proc
      .input(z.object({ req: z.string(), params: z.any() }))
      .output(z.any())
      .mutation(async ({ input }) => {
        const dtSettings = await getDevToolsConfig();
        if (!dtSettings.enabled) {
          throw new TRPCError({
            code: "PRECONDITION_FAILED",
            message: "Dev tools not enabled",
          });
        }
        invariant(obsConnection, "no OBS connection");
        return await obsConnection.callArbitraryDoNotUseOrYouWillBeFired(
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          input.req as any,
          input.params,
        );
      }),
  }),
});
