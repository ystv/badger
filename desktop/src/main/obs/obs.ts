import OBSWebSocket, {
  OBSRequestTypes,
  OBSResponseTypes,
  OBSWebSocketError,
} from "obs-websocket-js";
import { OBSSettings, getOBSSettings, saveOBSSettings } from "../base/settings";
import { getLogger } from "../base/logging";
import { inspect } from "node:util";
import { IntegrationManager } from "../base/integrations";
import { MockOBSConnection } from "./obs.mock";

const logger = getLogger("obs");

/*
 * This file contains OBSConnection, a wrapper around obs-websocket-js that provides a higher level, more typesafe API.
 * In general, anything that requires more than one call to OBS should go in obsHelpers.ts instead.
 */

export interface Scene {
  sceneIndex: number;
  sceneName: string;
}

export enum OBSBoundsTypes {
  OBS_BOUNDS_NONE = "OBS_BOUNDS_NONE" /**< no bounds */,
  OBS_BOUNDS_STRETCH = "OBS_BOUNDS_STRETCH" /**< stretch (ignores base scale) */,
  OBS_BOUNDS_SCALE_INNER = "OBS_BOUNDS_SCALE_INNER" /**< scales to inner rectangle */,
  OBS_BOUNDS_SCALE_OUTER = "OBS_BOUNDS_SCALE_OUTER" /**< scales to outer rectangle */,
  OBS_BOUNDS_SCALE_TO_WIDTH = "OBS_BOUNDS_SCALE_TO_WIDTH" /**< scales to the width  */,
  OBS_BOUNDS_SCALE_TO_HEIGHT = "OBS_BOUNDS_SCALE_TO_HEIGHT" /**< scales to the height */,
  OBS_BOUNDS_MAX_ONLY = "OBS_BOUNDS_MAX_ONLY" /**< no scaling, maximum size only */,
}
export type OBSBoundsType = keyof typeof OBSBoundsTypes;

export interface SceneItemTransform {
  alignment: number;
  boundsAlignment: number;
  boundsHeight: number;
  boundsType: OBSBoundsType;
  boundsWidth: number;
  cropBottom: number;
  cropLeft: number;
  cropRight: number;
  cropTop: number;
  height: number;
  positionX: number;
  positionY: number;
  rotation: number;
  scaleX: number;
  scaleY: number;
  sourceHeight: number;
  sourceWidth: number;
  width: number;
}

export interface SceneItem {
  inputKind: string;
  isGroup: null;
  sceneItemBlendMode: string;
  sceneItemEnabled: boolean;
  sceneItemId: number;
  sceneItemIndex: number;
  sceneItemLocked: boolean;
  sceneItemTransform: SceneItemTransform;
  sourceName: string;
  sourceType: string;
}

export interface MediaSourceSettings {
  local_file: string;
  is_local_file: boolean;
  looping: boolean;
  restart_on_activate: boolean;
}

export interface InputSettingsResult<T = unknown> {
  inputKind: string;
  inputSettings: T;
}

interface FFMPEGSourceDefaultSettings {
  buffering_mb: number;
  clear_on_media_end: boolean;
  is_local_file: boolean;
  linear_alpha: boolean;
  looping: boolean;
  reconnect_delay_sec: number;
  restart_on_activate: boolean;
  speed_percent: number;
}

export type FFMPEGSourceSettings = Partial<FFMPEGSourceDefaultSettings> &
  (
    | {
        is_local_file: true;
        local_file: string;
      }
    | {
        is_local_file: false;
        input: string;
        input_format: string;
      }
  );

interface VLCSourceDefaultSettings {
  loop: boolean;
  network_caching: number;
  playback_behavior: "stop_restart" | "pause_unpause" | "always_play";
  shuffle: boolean;
  subtitle: number;
  subtitle_enable: number;
  track: number;
}

export interface VLCSourceSettings extends Partial<VLCSourceDefaultSettings> {
  playlist: Array<{
    value: string;
  }>;
}

export function castMediaSourceSettings(
  x: InputSettingsResult,
): x is InputSettingsResult<MediaSourceSettings> {
  return x.inputKind === "ffmpeg_source";
}

export interface OBSVideoSettings {
  /**
   * Numerator of the fractional FPS value
   */
  fpsNumerator: number;
  /**
   * Denominator of the fractional FPS value
   */
  fpsDenominator: number;
  /**
   * Width of the base (canvas) resolution in pixels
   */
  baseWidth: number;
  /**
   * Height of the base (canvas) resolution in pixels
   */
  baseHeight: number;
  /**
   * Width of the output resolution in pixels
   */
  outputWidth: number;
  /**
   * Height of the output resolution in pixels
   */
  outputHeight: number;
}

export interface InputBasicInfo {
  inputKind: string;
  unversionedInputKind: string;
  inputName: string;
}

export default class OBSConnection {
  private obs!: OBSWebSocket;
  private logger = getLogger("OBSConnection");
  private constructor() {}
  public static async create(
    obsHost: string,
    obsPassword: string,
    obsPort = 4455,
  ) {
    const obsConnection = new OBSConnection();
    const obs = new OBSWebSocket();
    obs.on("ConnectionError", (err) => {
      obsConnection.logger.error("OBS error: " + String(err));
    });
    obsConnection.logger.info(
      "Connecting to OBS at",
      `ws://${obsHost}:${obsPort}`,
    );
    await obs.connect(`ws://${obsHost}:${obsPort}`, obsPassword, {
      rpcVersion: 1,
    });
    obsConnection.obs = obs;
    return obsConnection;
  }

  public async createScene(name: string): Promise<void> {
    await this._call("CreateScene", { sceneName: name });
  }

  public async listScenes(): Promise<Scene[]> {
    return (await this._call("GetSceneList")).scenes as unknown as Scene[];
  }

  public async getSceneItems(scene: string): Promise<SceneItem[]> {
    return (await this._call("GetSceneItemList", { sceneName: scene }))
      .sceneItems as unknown as SceneItem[];
  }

  public async createMediaSourceInput(
    scene: string,
    inputName: string,
    path: string,
  ) {
    const res = await this._call("CreateInput", {
      sceneName: scene,
      inputName,
      inputKind: "ffmpeg_source",
      inputSettings: {
        local_file: path,
        is_local_file: true,
        looping: false,
        restart_on_activate: true,
      } satisfies FFMPEGSourceSettings,
    });
    return res.sceneItemId;
  }

  public async getSourceSettings(
    inputName: string,
  ): Promise<InputSettingsResult> {
    return await this._call("GetInputSettings", { inputName });
  }

  public async replaceMediaSourceInScene(
    scene: string,
    inputName: string,
    path: string,
  ) {
    await this._call("SetInputSettings", {
      inputName: inputName,
      inputSettings: {
        local_file: path,
        is_local_file: true,
        looping: false,
        restart_on_activate: true,
      },
    });
  }

  public async setSceneItemTransform(
    sceneName: string,
    sceneItemId: number,
    transform: Partial<SceneItemTransform>,
  ): Promise<void> {
    await this._call("SetSceneItemTransform", {
      sceneName,
      sceneItemId,
      sceneItemTransform: transform,
    });
  }

  public async removeSceneItem(sceneTitle: string, itemId: number) {
    await this._call("RemoveSceneItem", {
      sceneName: sceneTitle,
      sceneItemId: itemId,
    });
  }

  public async listSources(): Promise<InputBasicInfo[]> {
    return (await this._call("GetInputList"))
      .inputs as unknown as InputBasicInfo[];
  }

  public async getVideoSettings(): Promise<OBSVideoSettings> {
    return await this._call("GetVideoSettings");
  }

  public async ping() {
    return await this._call("GetVersion");
  }

  public async callArbitraryDoNotUseOrYouWillBeFired(
    req: keyof OBSRequestTypes,
    requestData?: OBSRequestTypes[keyof OBSRequestTypes],
  ): Promise<OBSResponseTypes[keyof OBSResponseTypes]> {
    return await this._call(req, requestData);
  }

  public async close() {
    await this.obs.disconnect();
  }

  public onClosed(cb: (reason: OBSWebSocketError) => void) {
    this.obs.on("ConnectionClosed", cb);
  }

  private async _call<K extends keyof OBSRequestTypes>(
    req: K,
    requestData?: OBSRequestTypes[K],
  ): Promise<OBSResponseTypes[K]> {
    this.logger.debug("->OBS " + req + " " + inspect(requestData));
    const r = await this.obs.call(req, requestData);
    this.logger.debug("<-OBS " + req + " " + inspect(r));
    return r;
  }
}

export const OBSIntegration = new IntegrationManager<
  OBSConnection,
  OBSSettings
>(
  "obs",
  async (settings, notifyClosed) => {
    if (process.env.__USE_MOCK_OBS) {
      logger.warn("Using mock OBS connection");
      return new MockOBSConnection() as unknown as OBSConnection;
    }
    const conn = await OBSConnection.create(
      settings.host,
      settings.password,
      settings.port,
    );
    await conn.ping();
    conn.onClosed((reason) => {
      notifyClosed(reason);
    });
    return conn;
  },
  getOBSSettings,
  saveOBSSettings,
);
