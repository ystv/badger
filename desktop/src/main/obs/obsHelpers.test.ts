import { vi, describe, test, expect, beforeEach, afterEach } from "vitest";
import { addOrReplaceMediaAsScene, findScenes, MediaType } from "./obsHelpers";
import { MockOBSConnection } from "./__mocks__/obs";

vi.mock("./obs");
vi.mock("../media/mediaManagement", () => ({
  getLocalMedia: () => [
    {
      mediaID: 1,
      path: "TEST_PATH",
    },
  ],
}));
vi.mock("../base/selectedShow", async () => {
  const { BehaviorSubject } = await import("rxjs");
  return {
    selectedShow: new BehaviorSubject({
      id: 1,
      name: "Test",
      start: new Date(),
      rundowns: [],
      continuityItems: [],
      version: 1,
      ytBroadcastID: null,
      ytStreamID: null,
    }),
  };
});

describe("addOrReplaceMediaAsScene", () => {
  const testMedia: MediaType = {
    id: 1,
    name: "Test.mp4",
    state: "Ready",
    path: "",
    durationSeconds: 15,
    rawPath: "",
    containerType: "continuityItem",
    containerId: 1,
    containerName: "Test Continuity",
    order: 1,
  };
  let mobs: MockOBSConnection;
  beforeEach(async () => {
    mobs = (await import("./obs"))
      .obsConnection as unknown as MockOBSConnection;
  });
  afterEach(() => {
    mobs._reset();
  });

  test("add with no scenes", async () => {
    const res = await addOrReplaceMediaAsScene(testMedia, "none");
    expect(res).toEqual({
      done: true,
      warnings: [],
    });
    expect(mobs.scenes).toMatchInlineSnapshot(`
      [
        {
          "name": "1 - Test Continuity [Continuity #1]",
          "sources": [
            {
              "inputKind": "ffmpeg_source",
              "inputName": "Badger Media 1",
              "sceneItemId": 1,
            },
          ],
        },
      ]
    `);
  });

  test("no change", async () => {
    mobs.scenes.push({
      name: "1 - Test Continuity [Continuity #1]",
      sources: [
        {
          inputKind: "ffmpeg_source",
          inputName: "Badger Media 1",
          sceneItemId: 1,
        },
      ],
    });
    const res = await addOrReplaceMediaAsScene(testMedia, "none");
    expect(res).toEqual({
      done: false,
      warnings: [],
    });
    expect(mobs.scenes).toMatchSnapshot();
  });

  test("add to existing empty scene", async () => {
    mobs.scenes.push({
      name: "1 - Test Continuity [Continuity #1]",
      sources: [],
    });
    const res = await addOrReplaceMediaAsScene(testMedia, "none");
    expect(res).toEqual({
      done: true,
      warnings: [],
    });
    expect(mobs.scenes[0].sources).toHaveLength(1);
    expect(mobs.scenes).toMatchSnapshot();
  });

  test("replace in pre-existing scene with new media ID", async () => {
    mobs.scenes.push({
      name: "1 - Test Continuity [Continuity #1]",
      sources: [
        {
          inputKind: "ffmpeg_source",
          inputName: "Badger Media 999",
          sceneItemId: 1,
        },
      ],
    });
    const res = await addOrReplaceMediaAsScene(testMedia, "none");
    expect(res).toMatchInlineSnapshot(`
      {
        "done": false,
        "promptReplace": "replace",
        "warnings": [
          "Scene 1 - Test Continuity [Continuity #1] has a pre-existing Badger source for a different media file.",
        ],
      }
    `);
    expect(mobs.scenes[0].sources[0].inputName).toBe("Badger Media 999");

    const res2 = await addOrReplaceMediaAsScene(testMedia, "replace");
    expect(res2).toEqual({
      done: true,
      warnings: [],
    });
    expect(mobs.scenes).toMatchSnapshot();
    expect(mobs.scenes[0].sources[0].inputName).toBe("Badger Media 1");
  });

  test("non-Badger sources present", async () => {
    mobs.scenes.push({
      name: "1 - Test Continuity [Continuity #1]",
      sources: [
        {
          inputKind: "ffmpeg_source",
          inputName: "Badger Media 999",
          sceneItemId: 1,
        },
        {
          inputKind: "text_ft2_source_v2",
          inputName: "Text",
          sceneItemId: 2,
        },
      ],
    });
    const res = await addOrReplaceMediaAsScene(testMedia, "none");
    expect(res).toMatchInlineSnapshot(`
      {
        "done": false,
        "promptReplace": "force",
        "warnings": [
          "Scene 1 - Test Continuity [Continuity #1] has non-Badger sources in it. Cowardly refusing to overwrite.",
        ],
      }
    `);

    const res2 = await addOrReplaceMediaAsScene(testMedia, "force");
    expect(res2).toEqual({
      done: true,
      warnings: [],
    });
    expect(mobs.scenes).toMatchSnapshot();
    expect(mobs.scenes[0].sources).toHaveLength(1);
    expect(mobs.sourceSettings).toMatchSnapshot();
  });
});

describe("findContinuityScenes", () => {
  let mobs: MockOBSConnection;
  beforeEach(async () => {
    mobs = (await import("./obs"))
      .obsConnection as unknown as MockOBSConnection;
  });
  afterEach(() => {
    mobs._reset();
  });
  test("empty", async () => {
    expect(await findScenes()).toHaveLength(0);
  });
  test("one with no media", async () => {
    mobs.scenes.push({
      name: "1 - Test [#1]",
      sources: [],
    });
    const res = await findScenes();
    expect(res).toHaveLength(1);
    expect(res).toMatchInlineSnapshot(`
      [
        {
          "itemId": 1,
          "sceneName": "1 - Test [#1]",
          "sources": [],
          "type": "rundownItem",
        },
      ]
    `);
  });
  test("one with media", async () => {
    mobs.scenes.push({
      name: "1 - Test [#1]",
      sources: [
        {
          inputName: "media",
          inputKind: "ffmpeg_source",
          sceneItemId: 1,
        },
      ],
    });
    const res = await findScenes();
    expect(res).toHaveLength(1);
    expect(res[0].sources).toHaveLength(1);
    expect(res[0].sources[0]).toMatchInlineSnapshot(`
      {
        "inputKind": "ffmpeg_source",
        "isGroup": null,
        "sceneItemBlendMode": "OBS_BLEND_NORMAL",
        "sceneItemEnabled": true,
        "sceneItemId": 1,
        "sceneItemIndex": 0,
        "sceneItemLocked": false,
        "sceneItemTransform": null,
        "sourceName": "media",
        "sourceType": "input",
      }
    `);
  });
});
