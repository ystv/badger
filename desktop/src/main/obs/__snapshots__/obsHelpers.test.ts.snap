// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`addOrReplaceMediaAsScene > add to existing empty scene 1`] = `
[
  {
    "name": "1 - Test Continuity [Continuity #1]",
    "sources": [
      {
        "inputKind": "ffmpeg_source",
        "inputName": "Badger Media 1",
        "sceneItemId": 1,
      },
    ],
  },
]
`;

exports[`addOrReplaceMediaAsScene > no change 1`] = `
[
  {
    "name": "1 - Test Continuity [Continuity #1]",
    "sources": [
      {
        "inputKind": "ffmpeg_source",
        "inputName": "Badger Media 1",
        "sceneItemId": 1,
      },
    ],
  },
]
`;

exports[`addOrReplaceMediaAsScene > non-Badger sources present 2`] = `
[
  {
    "name": "1 - Test Continuity [Continuity #1]",
    "sources": [
      {
        "inputKind": "ffmpeg_source",
        "inputName": "Badger Media 1",
        "sceneItemId": 1,
      },
    ],
  },
]
`;

exports[`addOrReplaceMediaAsScene > non-Badger sources present 3`] = `
Map {
  "Badger Media 1" => {
    "is_local_file": true,
    "local_file": "TEST_PATH",
    "looping": false,
    "restart_on_activate": false,
  },
}
`;

exports[`addOrReplaceMediaAsScene > replace in pre-existing scene with new media ID 2`] = `
[
  {
    "name": "1 - Test Continuity [Continuity #1]",
    "sources": [
      {
        "inputKind": "ffmpeg_source",
        "inputName": "Badger Media 1",
        "sceneItemId": 1,
      },
    ],
  },
]
`;
